File: backend/src/main/java/me/nawa/appointment/domain/AppointmentStatus.java
package me.nawa.appointment.domain;

/**
 * 약속의 진행 상태입니다.
 */
public enum AppointmentStatus {
    PAYMENT_PENDING,
    RECRUITING,
    /**
     * 정원이 모두 찬 상태.
     *
     * 시간으로는 도달하지 않습니다 — 참여 마감 시각을 없앤 뒤로 이 값은 정원
     * 충족만 뜻합니다. 정원이 차지 않은 약속은 {@code RECRUITING}인 채로 활동
     * 시작 시각을 맞습니다.
     */
    FULL,
    IN_PROGRESS,
    /**
     * 활동 종료 시각이 지났지만 방장이 아직 출석을 확정하지 않은 상태.
     *
     * 활동 종료 시각이 지난 IN_PROGRESS 약속을
     * {@link me.nawa.appointment.service.AppointmentLifecycleScheduler}가 이 값으로
     * 옮겨 DB에 저장합니다.
     */
    AWAITING_ATTENDANCE,
    COMPLETED,
    CANCELLED;

    public boolean canTransitionTo(AppointmentStatus nextStatus) {
        if (nextStatus == null) {
            return false;
        }

        return switch (this) {
            case PAYMENT_PENDING ->
                    nextStatus == RECRUITING || nextStatus == CANCELLED;
            // 활동 시작 시각이 되면 방장의 별도 확정 없이 바로 진행중으로
            // 전환한다. 정원이 차지 않은 약속은 FULL을 거치지 않으므로
            // RECRUITING에서 곧장 IN_PROGRESS로 간다 — 이 경로가 없으면 정원
            // 미달 약속이 활동 시작 뒤에도 RECRUITING에 남는다.
            case RECRUITING ->
                    nextStatus == FULL
                            || nextStatus == IN_PROGRESS
                            || nextStatus == CANCELLED;
            // 정원이 차서 FULL이 된 뒤 활동 시작 전에 참여 취소가 생기면
            // 빈자리를 다시 채울 수 있도록 RECRUITING으로 되돌아간다.
            case FULL ->
                    nextStatus == RECRUITING
                            || nextStatus == IN_PROGRESS
                            || nextStatus == CANCELLED;
            // 스케줄러가 출석 확정 대기로 옮기기 전 몇 초 사이에도 화면은 이미
            // 출석 확정을 열어 주므로, 완료로 가는 길을 함께 둔다.
            case IN_PROGRESS ->
                    nextStatus == AWAITING_ATTENDANCE
                            || nextStatus == COMPLETED;
            case AWAITING_ATTENDANCE -> nextStatus == COMPLETED;
            case COMPLETED, CANCELLED -> false;
        };
    }
}

File: frontend/src/shared/lib/appointmentStatus.ts
/**
 * 약속 진행 상태.
 *
 * 백엔드 `AppointmentStatus` enum과 1:1이고, 값 집합과 전이 규칙의 정본은
 * `backend/docs/APPOINTMENT_DEPOSIT_STATE_MACHINE.md`다.
 *
 * appointment(목록·상세·생성)와 journey(타임라인 응답의 `appointment`)가 함께 쓰므로
 * shared에 둔다. feature끼리 직접 import하지 않는다는 규칙(DEVELOPMENT_CONVENTION.md)
 * 아래에서 두 feature가 같은 값 집합을 보려면 여기 말고 둘 곳이 없다 — 한쪽에 두고
 * 다른 쪽이 문자열로 받으면, 없어진 값이 남아도 타입 검사가 잡지 못한다(#387).
 */
export const APPOINTMENT_STATUSES = [
  'PAYMENT_PENDING',
  'RECRUITING',
  /** 정원이 모두 참. 시간으로 도달하는 경로는 없다. */
  'FULL',
  'IN_PROGRESS',
  /**
   * 활동 종료 시각이 지났지만 방장이 아직 출석을 확정하지 않음. 다른 값과 똑같이
   * DB에 저장된다 — 스케줄러가 이 전이를 기록한다.
   */
  'AWAITING_ATTENDANCE',
  'COMPLETED',
  'CANCELLED',
] as const

export type AppointmentStatus = (typeof APPOINTMENT_STATUSES)[number]
